modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import PropTypes from 'prop-types';
+import React from 'react';
+
+import {Editor} from './Editor';
+
+import '../css/main.scss';
+
+const BalloonEditor = ({config = {}, contents, name, ...otherProps}) => {
+	const defaultExtraPlugins = 'balloontoolbar,floatingspace';
+
+	const getConfig = () => {
+		const extraPlugins = config.extraPlugins
+			? `${config.extraPlugins}`
+			: '';
+
+		return {
+			...config,
+			extraAllowedContent: '*',
+			extraPlugins: `${extraPlugins}${defaultExtraPlugins}`,
+		};
+	};
+
+	let initialCssClass;
+
+	return (
+		<Editor
+			config={getConfig()}
+			name={name}
+			onBeforeLoad={(CKEDITOR) => {
+				CKEDITOR.disableAutoInline = true;
+
+				initialCssClass = CKEDITOR.env.cssClass;
+
+				CKEDITOR.env.cssClass = `${CKEDITOR.env.cssClass} lfr-balloon-editor`;
+			}}
+			onDestroy={() => {
+				CKEDITOR.env.cssClass = initialCssClass;
+			}}
+			onInstanceReady={(event) => {
+				const editor = event.editor;
+
+				const balloonToolbars = editor.balloonToolbars;
+
+				balloonToolbars.create({
+					buttons:
+						'Bold,Italic,Underline,RemoveFormat,Link,NumberedList,BulletedList,JustifyLeft,JustifyCenter,JustifyRight,JustifyBlock,Anchor',
+					refresh(editor, path) {
+						const tags = new Set([
+							'h1',
+							'h2',
+							'h3',
+							'p',
+							'span',
+							'strong',
+							'li',
+						]);
+
+						let result = false;
+
+						tags.forEach((tag) => {
+							if (path.contains(tag)) {
+								result = true;
+							}
+						});
```

Comment:
The `Set` is good for `O(1)` insertion, deletion, membership testing, but you are using it here for iteration `O(n)` (via `forEach`) and then you have a `path.contains()` which is itself a linear scan, so that's quadratic.

I want my Obama gif back!

What kind of object is `path`?

Reply:
To explain my intent (and give you back your gif): I thought of using a `Set` because I want each element to be unique, and now I'm thinking that, since I'm "adding" those items myself and that nothing from "outside" is going to touch `tags`, I could probably use a regular `Array`.

When it comes to iteration, you're right, that's something I didn't really think about.

The `path` object is a [`CKEDITOR.dom.elementPath`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_dom_elementPath.html)(that's a link to the docs, but I guess you also want to have a look at the [source](https://github.com/ckeditor/ckeditor4/blob/eb2d11644a796cd13cd047cd55541430359f9317/core/dom/elementpath.js)), and the idea behind this iteration is the following:
If the `path` contains any of those "tags", then we should show this toolbar ... 

I actually changed [this part of the code](https://github.com/liferay-frontend/liferay-portal/blob/12dac565f1b763bf2966e5d391dadf03c05d341a/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js#L58-L82) locally, to be like this

```js
balloonToolbars.create({
    buttons: 'Bold,Italic,Underline,RemoveFormat,Link,NumberedList,BulletedList,JustifyLeft,JustifyCenter,JustifyRight,JustifyBlock,Anchor',
    cssSelector: '*',
});
```

(*It's not something I came up with, since it was already "like" this in the original "draft"; and unfortunately, I can't find that pull request, and I can't remember where it was ... but I think it doesn't really matter*)


The good news is that the result is exactly the same.  This also looks more "simple" even though I haven't analyzed what it's doing "under the hood", but it's probably better for now than my quadratic monster.

*Another thing I didn't say (and probably should have because it's not obvious), is that this will probably change. I'm still trying to figure out how to use this CKEditor plugin to behave like we want it to.*  

![](https://media.giphy.com/media/3o7TKCRsacaH9u7HcA/source.gif)